In [1]:
#import libraries
import pandas as pd
import numpy as np

In [3]:
# Load existing Criteo analytical results

incrementality_test = pd.read_csv(
    "../outputs/tables/incrementality_statistical_test.csv"
)

incrementality_summary = pd.read_csv(
    "../outputs/tables/incrementality_summary.csv"
)

segment_incrementality = pd.read_csv(
    "../outputs/tables/segment_incrementality_summary.csv"
)

targeting_evaluation = pd.read_csv(
    "../outputs/tables/customer_targeting_model_evaluation.csv"
)

print("Incrementality test:")
display(incrementality_test)

print("\nIncrementality summary:")
display(incrementality_summary)

print("\nSegment incrementality:")
display(segment_incrementality)

print("\nTargeting evaluation:")
display(targeting_evaluation)

Incrementality test:


,Metric,Value
0,Control Conversion Rate,1.937588e-03
1,Treatment Conversion Rate,3.089461e-03
2,Absolute Conversion Lift,1.151873e-03
3,Relative Conversion Lift,5.944881e-01
4,Z-Statistic,2.851653e+01
5,P-Value,7.308263e-179
6,95% CI Lower,1.084505e-03
7,95% CI Upper,1.219242e-03



Incrementality summary:


,Metric,Control,Treatment,Absolute Lift,Relative Lift
0,Conversion Rate,0.001615,0.003394,0.001778,1.100926
1,Visit Rate,0.039671,0.048616,0.008944,0.225461



Segment incrementality:


,feature,segment,segment_population,treatment_population,control_population,treatment_conversions,control_conversions,treatment_conversion_rate,control_conversion_rate,absolute_lift,relative_lift,z_statistic,p_value,ci_lower,ci_upper,adequate_sample_size,statistically_reliable,attractive_segment
0,f8,"(3.634, 3.911]",3539789,3023235,516554,34761,3851,0.011498,0.007455,0.004043,0.542278,25.851980,2.311491e-147,0.003779,0.004306,True,True,True
1,f8,"(3.911, 3.972]",10439803,8859420,1580383,1950,212,0.000220,0.000134,0.000086,0.640801,6.918268,4.571990e-12,0.000065,0.000106,True,True,False
2,f6,"(-31.431, -6.699]",3622644,3102297,520347,24872,2450,0.008017,0.004708,0.003309,0.702763,25.530426,9.058711e-144,0.003098,0.003520,True,True,True
3,f6,"(-6.699, -2.411]",4418921,3752509,666412,4793,599,0.001277,0.000899,0.000378,0.421025,8.154817,3.497098e-16,0.000298,0.000459,True,True,False
4,f6,"(-2.411, 0.294]",5938027,5027849,910178,7046,1014,0.001401,0.001114,0.000287,0.257908,6.851059,7.330523e-12,0.000211,0.000363,True,True,False
5,f0,"(12.615, 21.923]",6989796,5949663,1040133,32999,3558,0.005546,0.003421,0.002126,0.621404,27.729076,3.115772e-169,0.001999,0.002253,True,True,True
6,f0,"(21.923, 24.436]",3494900,2968163,526737,2384,321,0.000803,0.000609,0.000194,0.317975,4.660473,3.154837e-06,0.000120,0.000268,True,True,False
7,f0,"(24.436, 26.745]",3494896,2964829,530067,1328,184,0.000448,0.000347,0.000101,0.290361,3.250194,1.153264e-03,0.000045,0.000156,True,True,False
8,f2,"(8.213000000000001, 8.723]",10484694,8918460,1566234,34384,3810,0.003855,0.002433,0.001423,0.584887,27.258891,1.303809e-163,0.001336,0.001510,True,True,True
9,f2,"(8.723, 9.052]",3494898,2964195,530703,2327,253,0.000785,0.000477,0.000308,0.646723,7.615829,2.620052e-14,0.000241,0.000375,True,True,False



Targeting evaluation:


,Model,ROC_AUC,Log_Loss
0,Treatment,0.946775,0.014145
1,Control,0.940009,0.008812


In [4]:
# Identify the segments previously classified as attractive

attractive_segments = segment_incrementality[
    segment_incrementality["attractive_segment"] == True
].copy()

attractive_segments[
    [
        "feature",
        "segment",
        "segment_population",
        "treatment_conversion_rate",
        "control_conversion_rate",
        "absolute_lift",
        "relative_lift",
        "statistically_reliable"
    ]
]

,feature,segment,segment_population,treatment_conversion_rate,control_conversion_rate,absolute_lift,relative_lift,statistically_reliable
0,f8,"(3.634, 3.911]",3539789,0.011498,0.007455,0.004043,0.542278,True
2,f6,"(-31.431, -6.699]",3622644,0.008017,0.004708,0.003309,0.702763,True
5,f0,"(12.615, 21.923]",6989796,0.005546,0.003421,0.002126,0.621404,True
8,f2,"(8.213000000000001, 8.723]",10484694,0.003855,0.002433,0.001423,0.584887,True


In [5]:
# Rank attractive segments by absolute and relative incremental lift

strongest_absolute = attractive_segments.sort_values(
    "absolute_lift",
    ascending=False
).reset_index(drop=True)

strongest_relative = attractive_segments.sort_values(
    "relative_lift",
    ascending=False
).reset_index(drop=True)

print("Attractive segments ranked by absolute lift:")
display(
    strongest_absolute[
        [
            "feature",
            "segment",
            "absolute_lift",
            "relative_lift"
        ]
    ]
)

print("\nAttractive segments ranked by relative lift:")
display(
    strongest_relative[
        [
            "feature",
            "segment",
            "absolute_lift",
            "relative_lift"
        ]
    ]
)

Attractive segments ranked by absolute lift:


,feature,segment,absolute_lift,relative_lift
0,f8,"(3.634, 3.911]",0.004043,0.542278
1,f6,"(-31.431, -6.699]",0.003309,0.702763
2,f0,"(12.615, 21.923]",0.002126,0.621404
3,f2,"(8.213000000000001, 8.723]",0.001423,0.584887



Attractive segments ranked by relative lift:


,feature,segment,absolute_lift,relative_lift
0,f6,"(-31.431, -6.699]",0.003309,0.702763
1,f0,"(12.615, 21.923]",0.002126,0.621404
2,f2,"(8.213000000000001, 8.723]",0.001423,0.584887
3,f8,"(3.634, 3.911]",0.004043,0.542278


In [6]:
# Display the existing T-learner evaluation results

targeting_evaluation

,Model,ROC_AUC,Log_Loss
0,Treatment,0.946775,0.014145
1,Control,0.940009,0.008812


## Criteo Customer Targeting — Results Reference

The Criteo analysis provides three complementary findings: advertising exposure generates incremental conversions, certain customer segments respond more strongly, and incremental response is concentrated among a smaller group of high-uplift customers.

### Overall Incrementality

| Metric                    |        Result |
| ------------------------- | ------------: |
| Control conversion rate   |       0.1938% |
| Treatment conversion rate |       0.3089% |
| Absolute lift             |     0.1152 pp |
| Relative lift             |        59.45% |
| P-value                   | 7.31 × 10⁻¹⁷⁹ |

Treatment exposure produced a statistically significant positive incremental effect, with a 59.45% relative improvement in conversion over control.

### High-Response Segments

Four feature-defined segments were classified as attractive:

| Feature | Segment           | Relative Lift |
| ------- | ----------------- | ------------: |
| f8      | (3.634, 3.911]    |        54.23% |
| f6      | (-31.431, -6.699] |        70.28% |
| f0      | (12.615, 21.923]  |        62.14% |
| f2      | (8.213, 8.723]    |        58.49% |

The **f8 segment** had the strongest absolute lift at 0.4043 percentage points, while **f6** had the highest relative lift at 70.28%.

### Customer-Level Targeting

A T-learner was used to estimate customer-level incremental response and rank customers by predicted uplift. The **Top 10%** of customers achieved an observed lift of **0.005672 (0.57 percentage points)** and captured approximately **65% of observed conversions**.

| Targeting Group | Population Share | Observed Lift |
| --------------- | ---------------: | ------------: |
| Top 10%         |              10% |       0.57 pp |
| Top 20%         |              20% |       0.28 pp |
| Top 30%         |              30% |       0.12 pp |
| Bottom 50%      |              50% |       0.15 pp |

The Top 10% showed approximately **3.8× higher observed uplift than the Bottom 50%**. Because uplift was not perfectly monotonic across all groups, the results support prioritizing the highest-uplift customers while continuing to validate targeting performance through controlled experimentation.

### Recommendation Evidence

Overall, the analysis supports three conclusions:

1. **Exposure works:** treatment generated a statistically significant incremental effect.
2. **Response varies across segments:** four feature-defined segments showed particularly strong performance.
3. **Incremental response is concentrated:** the highest-uplift customers produced the strongest observed targeting results.

In [7]:
# Existing targeting-strategy results from the Criteo evaluation

targeting_comparison = pd.DataFrame({
    "Target Group": [
        "Top 10%",
        "Top 20%",
        "Top 30%",
        "Random / Everyone"
    ],
    "Population %": [
        10.0,
        20.0,
        30.0,
        100.0
    ],
    "Observed Lift": [
        0.005672,
        0.002765,
        0.001223,
        0.001425
    ],
    "Recommended Action": [
        "Prioritize",
        "Consider if additional reach is required",
        "Deprioritize",
        "Baseline"
    ]
})

targeting_comparison

,Target Group,Population %,Observed Lift,Recommended Action
0,Top 10%,10.0,0.005672,Prioritize
1,Top 20%,20.0,0.002765,Consider if additional reach is required
2,Top 30%,30.0,0.001223,Deprioritize
3,Random / Everyone,100.0,0.001425,Baseline


In [8]:
# Compare Top 10% targeting with the random-targeting baseline

top_10_lift = targeting_comparison.loc[
    targeting_comparison["Target Group"] == "Top 10%",
    "Observed Lift"
].iloc[0]

random_lift = targeting_comparison.loc[
    targeting_comparison["Target Group"] == "Random / Everyone",
    "Observed Lift"
].iloc[0]

top_10_vs_random = top_10_lift / random_lift

print(f"Top 10% observed lift: {top_10_lift:.6f}")
print(f"Random observed lift: {random_lift:.6f}")
print(f"Top 10% / Random lift ratio: {top_10_vs_random:.2f}x")

Top 10% observed lift: 0.005672
Random observed lift: 0.001425
Top 10% / Random lift ratio: 3.98x


In [9]:
# Final targeting strategy comparison

targeting_comparison = pd.DataFrame({
    "Targeting Group": [
        "Top 10%",
        "Top 20%",
        "Top 30%",
        "Bottom 50%"
    ],
    "Population Share": [
        "10%",
        "20%",
        "30%",
        "50%"
    ],
    "Observed Lift": [
        0.005672,
        0.002765,
        0.001223,
        0.001500
    ]
})

targeting_comparison["Observed Lift (pp)"] = (
    targeting_comparison["Observed Lift"] * 100
)

targeting_comparison

,Targeting Group,Population Share,Observed Lift,Observed Lift (pp)
0,Top 10%,10%,0.005672,0.5672
1,Top 20%,20%,0.002765,0.2765
2,Top 30%,30%,0.001223,0.1223
3,Bottom 50%,50%,0.001500,0.1500


In [10]:
# Criteo targeting recommendation summary

criteo_recommendations = pd.DataFrame({
    "Area": [
        "Overall exposure",
        "Feature segment",
        "Feature segment",
        "Feature segment",
        "Feature segment",
        "Customer targeting",
        "Customer targeting",
        "Customer targeting"
    ],
    "Finding": [
        "Exposure generates incremental conversions",
        "f8 (3.634, 3.911]",
        "f6 (-31.431, -6.699]",
        "f0 (12.615, 21.923]",
        "f2 (8.213, 8.723]",
        "Top 10%",
        "Top 20%",
        "Top 30%"
    ],
    "Evidence": [
        "59.45% relative lift; p < 0.001",
        "0.4043 pp absolute lift; 54.23% relative lift",
        "0.3309 pp absolute lift; 70.28% relative lift",
        "0.2126 pp absolute lift; 62.14% relative lift",
        "0.1423 pp absolute lift; 58.49% relative lift",
        "0.5672 pp observed lift; ~65% of conversions",
        "0.2765 pp observed lift",
        "0.1223 pp observed lift"
    ],
    "Recommendation": [
        "Prioritize offer exposure where incremental response is expected",
        "Prioritize",
        "Prioritize",
        "Prioritize",
        "Prioritize",
        "Primary targeting group",
        "Secondary targeting group when more reach is needed",
        "Deprioritize for maximizing incremental response"
    ]
})

criteo_recommendations

,Area,Finding,Evidence,Recommendation
0,Overall exposure,Exposure generates incremental conversions,59.45% relative lift; p < 0.001,Prioritize offer exposure where incremental re...
1,Feature segment,"f8 (3.634, 3.911]",0.4043 pp absolute lift; 54.23% relative lift,Prioritize
2,Feature segment,"f6 (-31.431, -6.699]",0.3309 pp absolute lift; 70.28% relative lift,Prioritize
3,Feature segment,"f0 (12.615, 21.923]",0.2126 pp absolute lift; 62.14% relative lift,Prioritize
4,Feature segment,"f2 (8.213, 8.723]",0.1423 pp absolute lift; 58.49% relative lift,Prioritize
5,Customer targeting,Top 10%,0.5672 pp observed lift; ~65% of conversions,Primary targeting group
6,Customer targeting,Top 20%,0.2765 pp observed lift,Secondary targeting group when more reach is n...
7,Customer targeting,Top 30%,0.1223 pp observed lift,Deprioritize for maximizing incremental response


In [11]:
# Load the two Yelp merchant-level datasets used for Step 20.2

merchant_engagement = pd.read_csv(
    "../data/samples/yelp_merchant_engagement.csv"
)

merchant_priority = pd.read_csv(
    "../data/samples/yelp_merchant_priority_final.csv"
)

print("Merchant engagement table:")
print("Shape:", merchant_engagement.shape)
print("Columns:", merchant_engagement.columns.tolist())

print("\nMerchant priority table:")
print("Shape:", merchant_priority.shape)
print("Columns:", merchant_priority.columns.tolist())

Merchant engagement table:
Shape: (8812, 34)
Columns: ['business_id', 'name', 'address', 'city', 'state', 'postal_code', 'latitude', 'longitude', 'stars', 'review_count', 'is_open', 'attributes', 'categories', 'hours', 'total_reviews', 'earlier_reviews', 'recent_reviews', 'review_growth', 'average_rating', 'earlier_average_rating', 'recent_average_rating', 'rating_change', 'total_checkins', 'earlier_checkins', 'recent_checkins', 'checkin_growth', 'review_change', 'checkin_change', 'rating_change_for_score', 'review_trend_norm', 'checkin_trend_norm', 'rating_trend_norm', 'engagement_score', 'merchant_status']

Merchant priority table:
Shape: (300, 40)
Columns: ['business_id', 'name', 'address', 'city', 'state', 'postal_code', 'latitude', 'longitude', 'stars', 'review_count', 'is_open', 'attributes', 'categories', 'hours', 'total_reviews', 'earlier_reviews', 'recent_reviews', 'review_growth', 'average_rating', 'earlier_average_rating', 'recent_average_rating', 'rating_change', 'total_che

In [12]:
# Merchant health tier counts

merchant_status_counts = (
    merchant_engagement["merchant_status"]
    .value_counts()
    .rename_axis("merchant_status")
    .reset_index(name="merchant_count")
)

merchant_status_counts["percentage"] = (
    merchant_status_counts["merchant_count"]
    / len(merchant_engagement)
    * 100
)

merchant_status_counts

,merchant_status,merchant_count,percentage
0,Stable,4266,48.411257
1,Declining,2361,26.793010
2,Growing,2185,24.795733


In [13]:
# Validate that all 8,812 merchants are assigned a health status

print("Total merchants:", len(merchant_engagement))
print("Status counts sum:", merchant_status_counts["merchant_count"].sum())
print(
    "Missing merchant_status:",
    merchant_engagement["merchant_status"].isna().sum()
)

Total merchants: 8812
Status counts sum: 8812
Missing merchant_status: 0


In [14]:
# Priority-group counts in the 300-merchant sentiment sample

priority_counts = (
    merchant_priority["priority_group"]
    .value_counts()
    .rename_axis("priority_group")
    .reset_index(name="merchant_count")
)

priority_counts["percentage"] = (
    priority_counts["merchant_count"]
    / len(merchant_priority)
    * 100
)

priority_counts

,priority_group,merchant_count,percentage
0,Expand,147,49.000000
1,Growth Opportunity,79,26.333333
2,Monitor / Intervene,53,17.666667
3,Reassess,21,7.000000


In [15]:
# Inspect Monitor / Intervene merchants

monitor_intervene = merchant_priority[
    merchant_priority["priority_group"] == "Monitor / Intervene"
].copy()

monitor_intervene[
    [
        "business_id",
        "name",
        "city",
        "merchant_status",
        "engagement_score",
        "merchant_sentiment_score",
        "engagement_polarity",
        "sentiment_polarity",
        "priority_group"
    ]
].sort_values(
    "engagement_score",
    ascending=False
)

,business_id,name,city,merchant_status,engagement_score,merchant_sentiment_score,engagement_polarity,sentiment_polarity,priority_group
183,8hKi4SyepMGe2W1L4aQ_JQ,Italian Kitchen,Philadelphia,Growing,0.628968,-5.000000e-02,High,Negative,Monitor / Intervene
52,g-mnsod42xN2y_LdYpoORw,The Freshworks - Mayfair Store,Philadelphia,Growing,0.618437,-3.250000e-01,High,Negative,Monitor / Intervene
274,x4855nj7vfdXeVpg5fREZA,1100 Social,Philadelphia,Growing,0.601190,0.000000e+00,High,Negative,Monitor / Intervene
32,38yHZCHWgPZlDj5SqdNcpA,Mama Rosa's Pizza,Tampa,Growing,0.588523,-1.666667e-01,High,Negative,Monitor / Intervene
138,SjCxvaC-OZLnOqAqHTXG_Q,My Thai Restaurant,Philadelphia,Growing,0.573413,-3.500000e-01,High,Negative,Monitor / Intervene
25,SwBhaxfQPbyhsi0QHUAN0A,Seminole Hard Rock Hotel & Casino Tampa,Tampa,Growing,0.563718,-2.900000e-01,High,Negative,Monitor / Intervene
191,-AWclhh1_2VnqPylPgBU3g,Outback Steakhouse,Tampa,Growing,0.555952,-2.860000e-01,High,Negative,Monitor / Intervene
57,iZNR8-rqsBL2afDk4Zxe8A,Koch's Deli,Philadelphia,Growing,0.552198,-9.251859e-18,High,Negative,Monitor / Intervene
254,Ljtd29uAWCaGK046QuvPhQ,Popeyes Louisiana Kitchen,Tampa,Growing,0.549908,-5.542857e-01,High,Negative,Monitor / Intervene
16,wCluBbW9nzS7MEMFltMwJw,Tasties,Philadelphia,Growing,0.549908,-7.333333e-01,High,Negative,Monitor / Intervene


In [16]:
# Inspect Growth Opportunity merchants

growth_opportunity = merchant_priority[
    merchant_priority["priority_group"] == "Growth Opportunity"
].copy()

growth_opportunity[
    [
        "business_id",
        "name",
        "city",
        "merchant_status",
        "engagement_score",
        "merchant_sentiment_score",
        "engagement_polarity",
        "sentiment_polarity",
        "priority_group"
    ]
].sort_values(
    "engagement_score",
    ascending=True
)

,business_id,name,city,merchant_status,engagement_score,merchant_sentiment_score,engagement_polarity,sentiment_polarity,priority_group
285,jh-Sg2S4wlelkifvhjwSlg,China Garden,Philadelphia,Declining,0.420635,0.050000,Low,Positive,Growth Opportunity
240,1XQWvgGDqzDoEOg0_7eWKA,Six,Tampa,Declining,0.436676,0.350000,Low,Positive,Growth Opportunity
299,ny1N_Py01kVoGOvL9oKjMQ,Cask Social Kitchen,Tampa,Declining,0.441764,0.380000,Low,Positive,Growth Opportunity
277,1OfhM-ZKvcpxyxptCCzEwA,Independence Beer Garden,Philadelphia,Declining,0.443350,0.510000,Low,Positive,Growth Opportunity
80,_UOg5_pk9IhKee91eWrT4A,Sweet Lucy's Smokehouse,Philadelphia,Declining,0.448962,0.475000,Low,Positive,Growth Opportunity
...,...,...,...,...,...,...,...,...,...
117,gnpyyeWQ53WlDMJEq-YA_w,Tinto,Philadelphia,Declining,0.512897,0.490000,Low,Positive,Growth Opportunity
225,A37C-01zpQty9FmxvZrxEg,El Taco,Philadelphia,Declining,0.512922,0.720000,Low,Positive,Growth Opportunity
105,vXQLdqB-stL2JTP5AsPVaA,Samurai Blue Sushi & Sake Bar,Tampa,Declining,0.513045,0.019000,Low,Positive,Growth Opportunity
42,ktM6aYuzqWd-kA8SIWswiQ,Winnie's Manayunk,Philadelphia,Declining,0.513435,0.285000,Low,Positive,Growth Opportunity


In [17]:
# Selected merchant examples for final business recommendations

merchant_examples = pd.DataFrame([
    {
        "priority_group": "Monitor / Intervene",
        "merchant": "The Freshworks - Mayfair Store",
        "city": "Philadelphia",
        "engagement_score": 0.618437,
        "sentiment_score": -0.325,
        "business_interpretation": "High engagement but negative sentiment"
    },
    {
        "priority_group": "Monitor / Intervene",
        "merchant": "Tasties",
        "city": "Philadelphia",
        "engagement_score": 0.549908,
        "sentiment_score": -0.733333,
        "business_interpretation": "High engagement but strongly negative sentiment"
    },
    {
        "priority_group": "Monitor / Intervene",
        "merchant": "Boston Market",
        "city": "Philadelphia",
        "engagement_score": 0.547619,
        "sentiment_score": -0.800000,
        "business_interpretation": "High engagement but very strongly negative sentiment"
    },
    {
        "priority_group": "Growth Opportunity",
        "merchant": "China Garden",
        "city": "Philadelphia",
        "engagement_score": 0.420635,
        "sentiment_score": 0.050000,
        "business_interpretation": "Low engagement but positive sentiment"
    },
    {
        "priority_group": "Growth Opportunity",
        "merchant": "Cask Social Kitchen",
        "city": "Tampa",
        "engagement_score": 0.441764,
        "sentiment_score": 0.380000,
        "business_interpretation": "Low engagement with positive sentiment"
    },
    {
        "priority_group": "Growth Opportunity",
        "merchant": "Independence Beer Garden",
        "city": "Philadelphia",
        "engagement_score": 0.443350,
        "sentiment_score": 0.510000,
        "business_interpretation": "Low engagement with strong positive sentiment"
    }
])

merchant_examples

,priority_group,merchant,city,engagement_score,sentiment_score,business_interpretation
0,Monitor / Intervene,The Freshworks - Mayfair Store,Philadelphia,0.618437,-0.325000,High engagement but negative sentiment
1,Monitor / Intervene,Tasties,Philadelphia,0.549908,-0.733333,High engagement but strongly negative sentiment
2,Monitor / Intervene,Boston Market,Philadelphia,0.547619,-0.800000,High engagement but very strongly negative sen...
3,Growth Opportunity,China Garden,Philadelphia,0.420635,0.050000,Low engagement but positive sentiment
4,Growth Opportunity,Cask Social Kitchen,Tampa,0.441764,0.380000,Low engagement with positive sentiment
5,Growth Opportunity,Independence Beer Garden,Philadelphia,0.443350,0.510000,Low engagement with strong positive sentiment


## 20.2 Merchant Recommendations

### Merchant Health Across the Portfolio

The full merchant portfolio contains **8,812 merchants**: 4,266 Stable (48.41%), 2,185 Growing (24.80%), and 2,361 Declining (26.79%). More than one-quarter of merchants show declining engagement and may require additional attention.

### Priority-Based Recommendations

The 300-merchant sentiment sample combines engagement and customer sentiment to identify four priority groups:

| Priority Group      | Merchants |  Share |
| ------------------- | --------: | -----: |
| Expand              |       147 | 49.00% |
| Growth Opportunity  |        79 | 26.33% |
| Monitor / Intervene |        53 | 17.67% |
| Reassess            |        21 |  7.00% |

### Recommended Actions

* **Monitor / Intervene:** Focus on service quality and customer experience. These merchants have relatively strong engagement but negative sentiment, so the priority is to protect existing engagement while addressing recurring CX issues.

* **Growth Opportunity:** Focus on increasing engagement through targeted promotions, loyalty initiatives, and campaigns. Positive sentiment provides a favorable foundation for increasing customer activity.

* **Expand:** Continue investment where both engagement and sentiment support further growth.

* **Reassess:** Review merchant performance and customer feedback before allocating significant additional resources.

Overall, the recommendations distinguish between **protecting existing engagement through CX improvement** and **building engagement where customer sentiment is already favorable**, allowing merchant resources to be allocated according to the underlying risk and opportunity.